### PR TITLE
feat: Created Pubsublite-To-Bigquery Python Template and added Documentations

### PR DESCRIPTION
**Pull request by arindam-google**
- feat: Issue #653 
- Added all the files for the PubsubLite to Bigquery python dataproc template.
- Created the documentation for the PSL to BQ python dataproc template.

### CODE REVIEW

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            SparkSession
+                .builder
+                .appName("read-app")
+                .master("yarn")
+                .getOrCreate())
+
+        # Read
+        input_data=(spark.readStream \
+            .format(constants.FORMAT_PUBSUBLITE) \
+            .option(f"{constants.FORMAT_PUBSUBLITE}.subscription",f"projects/617357862702/locations/us-west1/subscriptions/{input_subscription}",) \
```
Comment (mugdhapattnaik) [inline python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:157]: The project ID variable containing the pubsublite subscription should be used here instead of hardcoding the ID.

Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py]: Removed the hardcoded values

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 2]
```diff
@@ -154,7 +162,7 @@ def run(self, spark: SparkSession, args: Dict[str, Any]) -> None:
         # Read
         input_data=(spark.readStream \
             .format(constants.FORMAT_PUBSUBLITE) \
-            .option(f"{constants.FORMAT_PUBSUBLITE}.subscription",f"projects/617357862702/locations/us-west1/subscriptions/{input_subscription}",) \
+            .option(f"{constants.FORMAT_PUBSUBLITE}.subscription",f"projects/{input_project_id}/locations/{region}/subscriptions/{input_subscription}",) \
```
Comment (saumyasinha-google) [inline python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:165]: Rather than passing multiple parameters for REGION we can simply fetch it via os.getenv

"projects/{input_project_id}/locations/{os.getenv('REGION')}/subscriptions/{input_subscription}"

**Review by saumyasinha-google — CHANGES_REQUESTED** [review]

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        output_dataset: str = args[constants.PUBSUBLITE_TO_BQ_OUTPUT_DATASET]
+        output_table: str = args[constants.PUBSUBLITE_TO_BQ_OUTPUT_TABLE]
+        pubsublite_checkpoint_location: str = args[constants.PUBSUBLITE_CHECKPOINT_LOCATION]
+        bq_temp_bucket: str = args[constants.PUBSUBLITE_TO_BQ_TEMPORARY_BUCKET]
+
+        logger.info(
+            "Starting Pubsublite to Bigquery spark job with parameters:\n"
+            f"{pprint.pformat(args)}"
+        )
+        # Set configuration to connect to Cassandra by overwriting the spark session
```
Comment (saumyasinha-google) [inline python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:154]: This should not be Cassandra.

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py (hunk 2 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:161]: Used os.environ.get instead of passing additional parameter for REGION. Changes are [here](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/e0ebb809e805d4ef75bbf9aa50c8454bb1d99ffd)

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py (hunk 3 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py]: Renamed Cassandra to Pubsublite

**Review by saumyasinha-google — DISMISSED** [review]

File: .gitignore [hunk 4]
```diff
@@ -24,4 +24,6 @@ dist
 pipeline.json
 .ipynb_checkpoints/
 .venv/
-metastore_db/
\ No newline at end of file
+metastore_db/
+
+psltobq.sh
```
Comment (vanshaj-bhatia) [inline .gitignore:29]: Why is this added?

File: python/dataproc_templates/pubsublite/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+Template for reading files from Cloud Pub/sub Lite and writing them to a BigQuery table.
+
+It uses the [Pubsublite Spark SQL Streaming](https://central.sonatype.com/artifact/com.google.cloud/pubsublite-spark-sql-streaming/1.0.0) to read the streaming data from Pubsublite and [Spark BigQuery connector](https://cloud.google.com/dataproc-serverless/docs/guides/bigquery-connector-spark-example) for writing to BigQuery
+
+## Arguments
+
+* `pubsublite.to.bq.input.subscription`: Cloud Storage location of the input files (format: `gs://bucket/...`)
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:7]: Description isn't correct.

File: python/dataproc_templates/pubsublite/README.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+* `pubsublite.to.bq.write.mode`: Output write mode (one of: append,overwrite,ignore,errorifexists)(Defaults to append)
+* `pubsublite.to.bq.output.dataset`: BigQuery output dataset name
+* `pubsublite.to.bq.output.table`: BigQuery output table name
+* `pubsublite.to.bq.bucket.name`: Temporary bucket for the Spark BigQuery connector
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:11]: If it is temp bucket, please add temp in the templateProperty name

File: python/dataproc_templates/pubsublite/README.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+## Usage
+
+```
+$ python main.py --template PUBSUBLITETOBQ --help
+usage: main.py --template PUBSUBTOBQ [-h] \
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:17]: Replace PUBSUBTOBQ with PUBSUBLITETOBQ

File: python/dataproc_templates/pubsublite/README.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    --pubsublite.to.bq.output.table PUBSUBLITE.BIGQUERY.OUTPUT.TABLE \
+    --pubsublite.to.bq.bucket.name PUBSUBLITE.BIGQUERY.BUCKET.NAME \
+    [--pubsublite.to.bq.write.mode {overwrite,append,ignore,errorifexists}] \
+    --pubsublite.to.bq.checkpoint.location PUBSUBLITE.BIGQUERY.CHECKPOINT.LOCATION
+```
+
+## Required JAR files
+
+This template requires the [Pubsublite Spark SQL Streaming](https://central.sonatype.com/artifact/com.google.cloud/pubsublite-spark-sql-streaming/1.0.0) to read the streaming data from Pubsublite and [Spark BigQuery connector](https://cloud.google.com/dataproc-serverless/docs/guides/bigquery-connector-spark-example) for writing to BigQuery.
+
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:30]: Repetetive

File: python/dataproc_templates/pubsublite/README.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+```
+$ python main.py --template PUBSUBLITETOBQ --help
+usage: main.py --template PUBSUBTOBQ [-h] \
+    --pubsublite.to.bq.input.subscription PUBSUBLITE.BIGQUERY.INPUT.SUBSCRIPTION \
+    --pubsublite.to.bq.project.id PUBSUBLITE.BIGQUERY.OUTPUT.PROJECT.ID \
+    --pubsublite.to.bq.output.dataset PUBSUBLITE.BIGQUERY.OUTPUT.DATASET \
+    --pubsublite.to.bq.output.table PUBSUBLITE.BIGQUERY.OUTPUT.TABLE \
+    --pubsublite.to.bq.bucket.name PUBSUBLITE.BIGQUERY.BUCKET.NAME \
+    [--pubsublite.to.bq.write.mode {overwrite,append,ignore,errorifexists}] \
+    --pubsublite.to.bq.checkpoint.location PUBSUBLITE.BIGQUERY.CHECKPOINT.LOCATION
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:24]: Is this supported? if yes, please test the functionality as it was not tested in spike task.

File: python/dataproc_templates/pubsublite/README.md [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+## Number of streams that will read from Pub/Sub subscription in parallel
+pubsublitetotal.receivers=5
+## Project that contains the output table
+pubsublitebq.output.project.id=<pubsublite to bq output project id>
+## BigQuery output dataset
+pubsublitebq.output.dataset=<bq output dataset>
+## BigQuery output table
+pubsublitebq.output.table=<bq output table>
+## Number of records to be written per message to BigQuery
+pubsublitebq.batch.size=1000
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:72]: Where is it used?

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py (hunk 2 above)
Comment (vanshaj-bhatia) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:161]: @arindam-google Let us take the whole subscription string from templateProperty

File: python/dataproc_templates/pubsublite/README.md [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+## Project that contains the input Pub/Sub subscription to be read
+pubsubliteinput.project.id=<pubsublite project id>
+## PubSublite subscription name
+pubsubliteinput.subscription=<pubsublite subscription>
+## Stream timeout, for how long the subscription will be read
+pubsublitetimeout.ms=60000
+## Streaming duration, how often wil writes to BQ be triggered
+pubsublitestreaming.duration.seconds=15
+## Number of streams that will read from Pub/Sub subscription in parallel
+pubsublitetotal.receivers=5
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:64]: Where is it used?

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+        query = (input_data.writeStream \
+            .format(constants.FORMAT_BIGQUERY) \
+            .option("temporaryGcsBucket", bq_temp_bucket) \
+            .option("checkpointLocation", pubsublite_checkpoint_location) \
+            .option("table", f"{output_project_id}.{output_dataset}.{output_table}") \
+            .trigger(processingTime="1 second") \
+            .start())
+
+        # Wait 120 seconds (must be >= 60 seconds) to start receiving messages.
+        query.awaitTermination(120)
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:176]: Remove hardcoding

File: python/dataproc_templates/util/template_constants.py [hunk 13]
```diff
@@ -286,27 +287,16 @@
...
-
-#Kafka To GCS
-KAFKA_GCS_BOOTSTRAP_SERVERS='kafka.gcs.bootstrap.servers'
-KAFKA_GCS_OUTPUT_LOCATION='kafka.gcs.output.location.gcs.path'
-KAFKA_TOPIC='kafka.gcs.topic'
-KAFKA_GCS_OUTPUT_FORMAT='kafka.gcs.output.format'
-KAFKA_GCS_OUPUT_MODE='kafka.gcs.output.mode'
-KAFKA_GCS_TERMINATION_TIMEOUT='kafka.gcs.termination.timeout'
-KAFKA_STARTING_OFFSET='kafka.gcs.starting.offset'
-KAFKA_GCS_CHECKPOINT_LOCATION='kafka.gcs.checkpoint.location'
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/util/template_constants.py:312]: Why are we removing these?

File: python/main.py [hunk 14]
```diff
@@ -67,9 +68,8 @@
     TemplateName.CASSANDRATOBQ: CassandraToBQTemplate,
     TemplateName.CASSANDRATOGCS: CassandraToGCSTemplate,
     TemplateName.HIVEDDLEXTRACTOR: HiveDDLExtractorTemplate,
-    TemplateName.KAFKATOGCS: KafkaToGCSTemplate,
-    TemplateName.KAFKATOBQ: KafkaToBigQueryTemplate,
```
Comment (vanshaj-bhatia) [inline python/main.py:71]: Why are we removing these?

**Review by vanshaj-bhatia — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 4 above)
Comment (arindam-google) [reply .gitignore]: Removed this one

File: python/dataproc_templates/pubsublite/README.md (hunk 9 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md:24]: `**_'pubsublite.to.bq.**_checkpoint.location_**': 'gs://arindamsarkar-test/pubsubtogcs-**_checkpoint-location_**',_**`
 'pubsublite.to.bq.input.subscription': 'arindamsarkar-test-sub-lite',
 'pubsublite.to.bq.input.timeout.ms': 60000,
 'pubsublite.to.bq.input.topic': None,
 'pubsublite.to.bq.output.dataset': 'arindamsarkar_test',
 'pubsublite.to.bq.output.table': 'pubsubtobq',
 'pubsublite.to.bq.project.id': 'yadavaja-sandbox',
 'pubsublite.to.bq.streaming.duration.seconds': 15,
 'pubsublite.to.bq.total.receivers': 5,
 'pubsublite.to.bq.write.mode': 'overwrite'}
23/03/06 10:17:28 INFO StateStoreCoordinatorRef: Registered StateStoreCoordinator endpoint
`23/03/06 10:17:28 INFO ResolveWriteToStream: **_Checkpoint root gs://arindamsarkar-test/pubsubtogcs-checkpoint-location resolved to gs://arindamsarkar-test/pubsubtogcs-checkpoint-location._**`
23/03/06 10:17:28 WARN ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
`23/03/06 10:17:29 INFO MicroBatchExecution: Starting [id = 4c8219ed-abc7-4768-9cb9-1ae0e40154dc, runId = **_9d059bfd-6870-47aa-bbf0-4cb578006b52]. Use gs://arindamsarkar-test/pubsubtogcs-checkpoint-location to store the query checkpoint._**`

Highlighted some of the places in the logs where the checkpoint location is being used while executing the template. So, based on these it seems it should be supported.

File: python/dataproc_templates/pubsublite/README.md (hunk 5 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Fixed -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/90477ae3a3561666fa24d39b0427348de969c06c)

File: python/dataproc_templates/pubsublite/README.md (hunk 6 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Fixed -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/90477ae3a3561666fa24d39b0427348de969c06c)

File: python/dataproc_templates/pubsublite/README.md (hunk 7 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Fixed -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/90477ae3a3561666fa24d39b0427348de969c06c)

File: python/dataproc_templates/pubsublite/README.md (hunk 8 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Fixed -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/90477ae3a3561666fa24d39b0427348de969c06c)

File: python/dataproc_templates/pubsublite/README.md (hunk 10 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Removed this parameter and executed the template and it ran successfully. So, removed this parameter -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/90477ae3a3561666fa24d39b0427348de969c06c)

File: python/dataproc_templates/pubsublite/README.md (hunk 11 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Removed this parameter and executed the template and it ran successfully. So, removed this parameter -> [commit1](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/90477ae3a3561666fa24d39b0427348de969c06c) & [commit2](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/673feef0cdcdc813c06567a0ee922a0ef2e0de13)

File: python/dataproc_templates/util/template_constants.py (hunk 13 above)
Comment (arindam-google) [reply python/dataproc_templates/util/template_constants.py]: Added them back -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/07f0faa83ca4f22067963cc265194044e4aac4fa)

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py (hunk 2 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py]: Done -> [commit](https://github.com/GoogleCloudPlatform/dataproc-templates/pull/658/commits/7ee38cf6e35e0bd285f547a4c02142684e698386)

File: python/dataproc_templates/pubsublite/README.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+## Stream timeout, for how long the subscription will be read
+pubsublitetimeout.ms=120000
+## Streaming duration, how often wil writes to BQ be triggered
+pubsublitestreaming.duration.seconds=15
+## Project that contains the output table
+pubsublitebq.output.project.id=<pubsublite to bq output project id>
+## BigQuery output dataset
+pubsublitebq.output.dataset=<bq output dataset>
+## BigQuery output table
+pubsublitebq.output.table=<bq output table>
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/README.md:69]: Please correct the template property names

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+                .appName("read-app")
+                .master("yarn")
+                .getOrCreate())
+
+        # Read
+        input_data=(spark.readStream \
+            .format(constants.FORMAT_PUBSUBLITE) \
+            .option(f"{constants.FORMAT_PUBSUBLITE}.subscription", input_subscription,) \
+            .load())
+
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:147]: This isn't looking correct.

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 17]
```diff
@@ -0,0 +1,161 @@
...
+        query = (input_data.writeStream \
+            .format(constants.FORMAT_BIGQUERY) \
+            .option("temporaryGcsBucket", bq_temp_bucket) \
+            .option("checkpointLocation", pubsublite_checkpoint_location) \
+            .option("table", f"{output_project_id}.{output_dataset}.{output_table}") \
+            .trigger(processingTime="1 second") \
+            .start())
+
+        # Wait timeout_ms/1000 seconds (must be >= 60 seconds) to start receiving messages.
+        query.awaitTermination(timeout_ms/1000)
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:160]: Why are we doing this? Can't we specify to provide timeout in seconds?

Comment (vanshaj-bhatia) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py:160]: Please update example submission. I do not think this property is being specified there.

File: python/dataproc_templates/util/template_constants.py [hunk 18]
```diff
@@ -310,3 +310,17 @@
...
+PUBSUBLITE_TO_BQ_INPUT_TIMEOUT_MS = "pubsublite.to.bq.input.timeout.ms"
+PUBSUBLITE_TO_BQ_STREAMING_DURATION_SECONDS = "pubsublite.to.bq.streaming.duration.seconds"
+PUBSUBLITE_TO_BQ_WRITE_MODE = "pubsublite.to.bq.write.mode"
+PUBSUBLITE_TO_BQ_TOTAL_RECEIVERS = "pubsublite.to.bq.total.receivers"
+PUBSUBLITE_TO_BQ_PROJECT_ID = "pubsublite.to.bq.project.id"
+PUBSUBLITE_TO_BQ_OUTPUT_DATASET = "pubsublite.to.bq.output.dataset"
+PUBSUBLITE_TO_BQ_OUTPUT_TABLE = "pubsublite.to.bq.output.table"
+PUBSUBLITE_TO_BQ_BATCH_SIZE = "pubsublite.to.bq.batch.size"
+PUBSUBLITE_TO_BQ_TEMPORARY_BUCKET = "pubsublite.to.bq.temp.bucket.name"
+PUBSUBLITE_CHECKPOINT_LOCATION = "pubsublite.to.bq.checkpoint.location"
```
Comment (vanshaj-bhatia) [inline python/dataproc_templates/util/template_constants.py:326]: Update PUBSUBLITE_CHECKPOINT_LOCATION to resemble other template varaibles.

**Review by vanshaj-bhatia — CHANGES_REQUESTED** [review]

**Comment by vanshaj-bhatia** [comment]
Unit test cases are missing, please add the same.

**Comment by arindam-google** [comment]
> Unit test cases are missing, please add the same.

Added the unit test file for PSL to BQ python template

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py (hunk 17 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py]: Done

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+            default=120000,
+            help='Stream timeout, for how long the subscription will be read'
+        )
+        parser.add_argument(
+            f'--{constants.PUBSUBLITE_TO_BQ_STREAMING_DURATION_SECONDS}',
+            dest=constants.PUBSUBLITE_TO_BQ_STREAMING_DURATION_SECONDS,
+            required=False,
+            default=15,
+            help='Streaming duration, how often wil writes to BQ be triggered'
+        )
```
[the remark this answers is not in the document]
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py]: It was not used anywhere so removed it

File: python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+    """
+
+    @staticmethod
+    def parse_args(args: Optional[Sequence[str]] = None) -> Dict[str, Any]:
+        parser: argparse.ArgumentParser = argparse.ArgumentParser()
+
+        parser.add_argument(
+            f'--{constants.PUBSUBLITE_TO_BQ_INPUT_TOPIC}',
+            dest=constants.PUBSUBLITE_TO_BQ_INPUT_TOPIC,
+            required=False,
```
[the remark this answers is not in the document]
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/pubsublite_to_bigquery.py]: It was not used anywhere so removed it

File: python/dataproc_templates/pubsublite/README.md (hunk 15 above)
Comment (arindam-google) [reply python/dataproc_templates/pubsublite/README.md]: Done

**Comment by vanshaj-bhatia** [comment]
Closing this PR as of now, because PSL to BQ isn't working.
